File: checkers/127/main.cpp
#include <checkers/testlib.h>
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <vector>
#include <sstream>

using namespace NTestlib;

int main(int argc, char *argv[])
{
    InitInteractor(argc, argv);
    int n = File.ReadInt();
    std::cout << n << std::endl;
    std::vector<int> a;
    for (int i = 0; i < n; i++)
        a.push_back(File.ReadInt());
    int m = File.ReadInt();
    std::cout << m << std::endl;
    for (int i = 0; i < m; i++)
    {
        int cnt = 0;
        int val = File.ReadInt();
        std::cout << val << std::endl;
        int answer;
        answer = Ans.ReadInt();
        while (true)
        {
            if (cnt >= 50)
            {
                QuitWith(WA, "Too many requests");
            }
            char requestType = Out.ReadChar();
            bool isFind = false;
            switch (requestType)
            {
            case '!':
            {
                int userAnswer = Out.ReadInt();
                if (userAnswer != answer)
                {
                    std::stringstream msg;
                    msg << "Incorrect user output. User answer: " << userAnswer << ". Correct answer: " << answer;
                    QuitWith(WA, msg.str());
                }
                isFind = true;
                break;
            }
            case '?':
            {
                int userAnswer = Out.ReadInt(1, n);
                if (userAnswer > 0 && userAnswer <= n)
                    std::cout << a[userAnswer - 1] << std::endl;
                else
                {
                    std::stringstream msg;
                    msg << "Wrong index " << userAnswer;
                    QuitWith(PE, msg.str());
                }
                break;
            }
            default:
            {
                std::stringstream msg;
                msg << "Incorrect input command: " << requestType;
                QuitWith(PE, msg.str());
            }
            break;
            }
            Out.Match('\n');
            if (isFind)
            {
                break;
            }
            cnt++;
        }
    }
    QuitWith(AC, "OK");
    std::cerr << "INT\n";
    return 0;
}

File: checkers/127/main.go
package main

import (
	"fmt"
	"testlib"
	"testlib/core"
)

func main() {
	defer func() {
		testlib.GetStream(testlib.OutputToUser).Close()
		testlib.GetStream(testlib.InputFromUser).Close()
	}()
	inputStream := testlib.GetReader(testlib.TaskInput) // эквивалент
	outputStream := testlib.GetReader(testlib.TaskOutput)
	userOutputStream := testlib.GetWriter(testlib.OutputToUser)
	userInputStream := testlib.GetReader(testlib.InputFromUser)
	n := core.ReadInt(inputStream)
	var data []int
	core.WriteStringToWriter(userOutputStream, fmt.Sprintf("%d\n", n))
	core.FlushWriter(userOutputStream)
	for i := 0; i < n; i++ {
		data = append(data, core.ReadInt(inputStream))
	}
	m := core.ReadInt(inputStream)
	core.WriteStringToWriter(userOutputStream, fmt.Sprintf("%d\n", m))
	core.FlushWriter(userOutputStream)
	for i := 0; i < m; i++ {
		cnt := 0
		val := core.ReadInt(inputStream)
		core.WriteStringToWriter(userOutputStream, fmt.Sprintf("%d\n", val))
		core.FlushWriter(userOutputStream)
		answer := core.ReadInt(outputStream)
		for {
			if cnt >= 50 {
				testlib.WriteAnswer("Too many requests", testlib.WA)
				return
			}
			requestType := core.ReadByte(userInputStream)
			isFind := false
			switch requestType {
			case byte('!'):
				userAnswer := core.ReadInt(userInputStream)
				if userAnswer != answer {
					testlib.WriteAnswer(fmt.Sprintf("Incorrect user output. User answer: %d. Correct answer: %d", userAnswer, answer), testlib.WA)
					return
				}
				isFind = true
				break
			case byte('?'):
				userAnswer := core.ReadInt(userInputStream)
				if userAnswer > 0 && userAnswer <= n {
					core.WriteStringToWriter(userOutputStream, fmt.Sprintf("%d\n", data[userAnswer-1]))
					core.FlushWriter(userOutputStream)
				} else {
					testlib.WriteAnswer(fmt.Sprintf("Index is out of range. User index: %d. Data length: %d", userAnswer, len(data)), testlib.PE)
					return
				}
				break
			default:
				testlib.WriteAnswer(fmt.Sprintf("Invalid request type: %s", string(requestType)), testlib.PE)
				return
			}
			if isFind {
				break
			}
			cnt++
		}
	}
	testlib.WriteAnswer("OK", testlib.OK)
}
